### PR TITLE
v3 support

### PR DESCRIPTION
**Pull request by MegaByte875**
-  upgrade nebula client

-  update apis/template configuration

### CODE REVIEW

**Comment by wey-gu** [comment]
Dear @porscheme 
Here is the v3 support of the operator, if you are interested to try it ;)

Thanks.
BR//Wey

File: pkg/nebula/meta_client.go [hunk 1]
```diff
@@ -268,31 +300,31 @@ func (m *metaClient) BalanceLeader() error {
...
 			if err != nil {
 				return err
 			}
 			if resp.Code != nebula.ErrorCode_SUCCEEDED {
 				return errors.Errorf("retry balance code %d", resp.Code)
 			}
-			log.Info("balance plan running now")
-			return m.BalanceStatus(resp.Id)
+			log.Info("balance job running now")
+			return m.BalanceStatus(*resp.GetResult_().JobID, space)
```
Comment (kqzh) [inline pkg/nebula/meta_client.go:327]: It seems balanceStatus is a long task, do we need use async？

File: pkg/nebula/meta_client.go [hunk 2]
```diff
@@ -306,68 +338,78 @@ func (m *metaClient) balance(req *meta.BalanceReq) error {
...
+func (m *metaClient) BalanceStatus(jobID int32, space []byte) error {
+	statusFn := func(ctx context.Context) (bool, error) {
+		req := &meta.AdminJobReq{
+			Op:    meta.AdminJobOp_SHOW,
+			Cmd:   meta.AdminCmd_STATS,
+			Paras: [][]byte{[]byte(strconv.FormatInt(int64(jobID), 10)), space},
 		}
+		resp, err := m.client.RunAdminJob(req)
+		if err != nil {
+			return false, err
```
Comment (kqzh) [inline pkg/nebula/meta_client.go:383]: if meta done or leader changed? balanceStatus will never stop?

File: pkg/util/retry/retry.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	return wait.PollImmediateUntil(interval, func() (bool, error) {
+		done, err := fn(ctx)
+		if err != nil {
+			if done {
+				return false, err
+			}
+		} else if done {
+			return true, nil
+		}
+		return false, nil
```
Comment (kqzh) [inline pkg/util/retry/retry.go:48]: how about 
```go
done, err := fn(ctx)
if err != nil {
  return false, err
}
return done, nil
```

File: pkg/controller/component/storaged_cluster.go [hunk 4]
```diff
@@ -167,6 +183,36 @@ func (c *storagedCluster) syncStoragedConfigMap(nc *v1alpha1.NebulaCluster) (*co
...
+	endpoints := []string{nc.GetMetadThriftConnAddress()}
+	metaClient, err := nebula.NewMetaClient(endpoints)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := metaClient.Disconnect()
+		if err != nil {
+			return
+		}
```
Comment (veezhang) [inline pkg/controller/component/storaged_cluster.go:196]: `_ = metaClient.Disconnect()` ?

**Comment by veezhang** [comment]
How do you think about it when it's needed multiple versions at the same kubernetes?

File: pkg/controller/component/storaged_cluster.go [hunk 5]
```diff
@@ -123,6 +126,19 @@ func (c *storagedCluster) syncStoragedWorkload(nc *v1alpha1.NebulaCluster) error
 		return utilerrors.ReconcileErrorf("waiting for storaged cluster %s running", newWorkload.GetName())
 	}
 
+	oldReplicas := extender.GetReplicas(oldWorkload)
+	newReplicas := extender.GetReplicas(newWorkload)
+	if !nc.Status.Storaged.HostsAdded || *newReplicas > *oldReplicas {
+		if err := c.addStorageHosts(nc, *oldReplicas, *newReplicas); err != nil {
```
[the remark this answers is not in the document]
Comment (MegaByte875) [reply pkg/controller/component/storaged_cluster.go:132]: Here only executed once

File: pkg/controller/component/storaged_cluster.go [hunk 6]
```diff
@@ -123,6 +126,19 @@ func (c *storagedCluster) syncStoragedWorkload(nc *v1alpha1.NebulaCluster) error
...
+			_, ok := err.(*net.DNSError)
+			if ok {
+				return nil
```
[the remark this answers is not in the document]
Comment (MegaByte875) [reply pkg/controller/component/storaged_cluster.go:135]: Here capture DNSError will not block subsequent reconcile logic

File: pkg/nebula/meta_client.go (hunk 1 above)
Comment (MegaByte875) [reply pkg/nebula/meta_client.go]: Yes, this's not a ideal implementation. I will save the last job in status field.

**Comment by porscheme** [comment]
> 

@MegaByte875 @wey-gu how I test this?

**Comment by wey-gu** [comment]
@MegaByte875 could you help @porscheme with a step by step guide?
I think to clone your branch should be ok but guides to help the build to deploy may help a lot to him/me.
Thanks!

**Comment by MegaByte875** [comment]
I will update helm charts later, you can deploy nebula v3 by helm @porscheme

File: pkg/controller/component/storaged_cluster.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (veezhang) [reply pkg/controller/component/storaged_cluster.go:132]: For example:

1. `addStorageHosts` successfully.
2. `syncComponentStatus` failed.
3. reconcile, and call `addStorageHosts` again.

File: pkg/controller/component/storaged_cluster.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (veezhang) [reply pkg/controller/component/storaged_cluster.go:135]: So, it's the `DNSError` a normal behavior?
If all the following succeed, then add host succeeded?

File: pkg/util/retry/retry.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	cx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	return Until(cx, interval, fn)
+}
+
+func Until(ctx context.Context, interval time.Duration, fn DoneFunc) error {
+	var stop <-chan struct{}
+	if ctx != nil {
+		stop = ctx.Done()
+	}
```
[the remark this answers is not in the document]
Comment (veezhang) [reply pkg/util/retry/retry.go:38]: How use `ctx.Done()` directly,  `ctx` generally cannot be nil.

**Review by veezhang — APPROVED** [review]
LGTM

**Comment by porscheme** [comment]
operator scheduler is crashing, see that attached
Capture
---my helm values file---
```
image:
  nebulaOperator:
    image: vesoft/nebula-operator:v1.0.0
    imagePullPolicy: Always
  kubeRBACProxy:
    image: gcr.io/kubebuilder/kube-rbac-proxy:v0.8.0
    imagePullPolicy: Always
  kubeScheduler:
    image: k8s.gcr.io/kube-scheduler:v1.18.8
    imagePullPolicy: Always

imagePullSecrets: []
kubernetesClusterDomain: ""

controllerManager:
  create: true
  replicas: 2
  env: []
  resources:
    limits:
      cpu: 200m
      memory: 200Mi
    requests:
      cpu: 100m
      memory: 100Mi

admissionWebhook:
  create: true

scheduler:
  create: true
  schedulerName: nebula-scheduler
  replicas: 1
  env: []
  resources:
    limits:
      cpu: 200m
      memory: 200Mi
    requests:
      cpu: 100m
      memory: 100Mi

# Enable openkruise scheme for controller manager. (default false)
enableKruise: false

# Period at which the controller forces the repopulation of its local object stores. (default 1h0m0s)
syncPeriod:

# Namespace the controller watches for updates to Kubernetes objects, If empty, all namespaces are watched.
watchNamespace:

# The address the metric endpoint binds to. (default ":8080")
metricsBindAddr: ":8080"

# The address the probe endpoint binds to. (default: ":8081")
healthProbeBindAddr: ":8081"

# The TCP port the Webhook server binds to. (default 9443)
webhookBindPort: 9443

# Maximum number of concurrently running reconcile loops for nebula cluster (default 3)
maxConcurrentReconciles: 3

nodeSelector:
  agentpool: metad

tolerations: []

affinity: {}
```

**Pull request merged**